apps/emqx_machine/src/emqx_cluster_rpc.erl

```diff
@@ -0,0 +1,325 @@
+        [] -> caught_up;
+        [#cluster_rpc_mfa{mfa = MFA}] -> {still_lagging, NextId, MFA}
+    end.
+
+do_catch_up(ToTnxId, Node) ->
+    case mnesia:wread({?CLUSTER_COMMIT, Node}) of
+        [] ->
+            commit(Node, ToTnxId),
+            caught_up;
+        [#cluster_rpc_commit{tnx_id = LastAppliedId}] when ToTnxId =:= LastAppliedId ->
+            caught_up;
+        [#cluster_rpc_commit{tnx_id = LastAppliedId}] when ToTnxId > LastAppliedId ->
+            CurTnxId = LastAppliedId + 1,
+            [#cluster_rpc_mfa{mfa = MFA}] = mnesia:read(?CLUSTER_MFA, CurTnxId),
+            case apply_mfa(CurTnxId, MFA) of
+                ok -> ok = commit(Node, CurTnxId);
+                {error, Reason} -> mnesia:abort(Reason);
+                Other -> mnesia:abort(Other)
+            end;
+        [#cluster_rpc_commit{tnx_id = LastAppliedId}]  ->
+            Reason = lists:flatten(io_lib:format("~p catch up failed by LastAppliedId(~p) > ToTnxId(~p)",
+                [Node, LastAppliedId, ToTnxId])),
+            ?LOG(error, Reason),
+            {error, Reason}
+    end.
+
+commit(Node, TnxId) ->
+    ok = mnesia:write(?CLUSTER_COMMIT, #cluster_rpc_commit{node = Node, tnx_id = TnxId}, write).
+
+get_latest_id() ->
+    case mnesia:last(?CLUSTER_MFA) of
+        '$end_of_table' -> 0;
+        Id -> Id
+    end.
+
+handle_mfa_write_event(#cluster_rpc_mfa{tnx_id = EventId, mfa = MFA}, Data) ->
+    #{node := Node, retry_interval := RetryMs} = Data,
+    {atomic, LastAppliedId} = transaction(fun get_last_applied_id/2, [Node, EventId - 1]),
+    if LastAppliedId + 1 =:= EventId ->
+        case apply_mfa(EventId, MFA) of
+            ok ->
+                case transaction(fun commit/2, [Node, EventId]) of
+                    {atomic, ok} ->
+                        {next_state, ?REALTIME, Data};
+                    _ -> {next_state, ?CATCH_UP, Data, [?CATCH_UP_AFTER(RetryMs)]}
+                end;
+            _ -> {next_state, ?CATCH_UP, Data, [?CATCH_UP_AFTER(RetryMs)]}
+        end;
+        LastAppliedId >= EventId -> %% It's means the initiator receive self event or other receive stale event.
+            keep_state_and_data;
+        true ->
+            ?LOG(error, "LastAppliedId+1<EventId, maybe the mnesia event'order is messed up! restart process:~p",
+                [{LastAppliedId, EventId, MFA, Node}]),
+            {stop, {"LastAppliedId+1<EventId", {LastAppliedId, EventId, MFA, Node}}}
+    end.
+
+get_last_applied_id(Node, Default) ->
+    case mnesia:wread({?CLUSTER_COMMIT, Node}) of
+        [#cluster_rpc_commit{tnx_id = TnxId}] -> TnxId;
+        [] ->
+            commit(Node, Default),
+            Default
+    end.
+
+init_mfa(Node, MFA) ->
+    mnesia:write_lock_table(?CLUSTER_MFA),
+    LatestId = get_latest_id(),
+    ok = do_catch_up_in_one_trans(LatestId, Node),
+    TnxId = LatestId + 1,
+    MFARec = #cluster_rpc_mfa{tnx_id = TnxId, mfa = MFA, initiator = Node, created_at = erlang:localtime()},
```

Comment:
I see a clock skew problem here:
1) If we want to guarantee total order of the operations in the cluster, then an atomic counter should be used here instead of local time
2) If we just want to guarantee partial ordering of operations, then it should use monotonic time

I would prefer the first alternative, since it has less surprises.

Reply:
Yes, we use the `TnxId = LatestId + 1 ` in the transaction(lock MFA table) to guarantee the atomic counter, the created_at is only for troubleshooting.

Reply:
Thanks for clarification